apps/cpp_rpc/rpc_server.cc

```diff
@@ -0,0 +1,363 @@
+          }
+
+          int status_second = 0;
+          waitpid_eintr(&status_second);
+
+          // Logging.
+          if (finished_first == timer_pid) {
+            LOG(INFO) << "Child pid=" << worker_pid << " killed (timeout = " << timeout
+                      << "), Process status =" << status_second;
+          } else if (finished_first == worker_pid) {
+            LOG(INFO) << "Child pid=" << timer_pid << " killed, Process status =" << status_second;
+          }
+        } else {
+          auto pid = fork();
+          if (pid == 0) {
+            ServerLoopProc(conn, addr);
+            exit(0);
+          }
+          // Wait for the result
+          int status = 0;
+          wait(&status);
+          LOG(INFO) << "Child pid=" << pid << " exited, Process status =" << status;
+        }
+      #else
+        // step 3: serving
+        std::future<void> proc(std::async(std::launch::async,
+                                          &RPCServer::ServerLoopProc, this, conn, addr));
+        // wait until server process finish or timeout
+        if (timeout) {
+          // Autoterminate after timeout
+          proc.wait_for(std::chrono::seconds(timeout));
+        } else {
+          // Wait for the result
+          proc.get();
+        }
+      #endif
+      // close from our side.
+      LOG(INFO) << "Socket Connection Closed";
+      conn.Close();
+    }
+  }
+
+
+  /*!
+   * \brief AcceptConnection Accepts the RPC Server connection.
+   * \param tracker Tracker details.
+   * \param conn New connection information.
+   * \param addr New connection address information.
+   * \param opts Parsed options for socket
+   * \param ping_period Timeout for select call waiting
+   */
+  void AcceptConnection(TrackerClient *tracker,
+                        common::TCPSocket *conn_sock,
+                        common::SockAddr *addr,
+                        std::string *opts,
+                        int ping_period = 2) {
+    std::set <std::string> old_keyset;
+    std::string matchkey;
+
+    // Report resource to tracker and get key
+    tracker->ReportResourceAndGetKey(my_port_, &matchkey);
+
+    while (1) {
+      tracker->WaitConnectionAndUpdateKey(listen_sock_, my_port_, ping_period, &matchkey);
+      common::TCPSocket conn = listen_sock_.Accept(addr);
+
+      int code = kRPCMagic;
+      CHECK_EQ(conn.RecvAll(&code, sizeof(code)), sizeof(code));
+      if (code != kRPCMagic) {
+        conn.Close();
+        LOG(FATAL) << "Client connected is not TVM RPC server";
+        continue;
+      }
+
+      int keylen = 0;
+      CHECK_EQ(conn.RecvAll(&keylen, sizeof(keylen)), sizeof(keylen));
+
+      #define CLIENT_HEADER "client:"
+      #define SERVER_HEADER "server:"
+
+      std::string expect_header = CLIENT_HEADER + matchkey;
+      std::string server_key = SERVER_HEADER + key_;
+      if (size_t(keylen) < expect_header.length()) {
+        conn.Close();
+        LOG(FATAL) << "Wrong client header length";
```

Comment:
We should not get to LOG(FATAL) when there is a bad connection(in the case of an attack), LOG(INFO) is fine to log the event